cpp/src/merge/merge.cu

```diff
@@ -353,76 +354,154 @@ std::unique_ptr<column> column_merger::operator()<cudf::dictionary32>(column_vie
-    v_merged_cols.reserve(n_cols);
+    //    
+    auto const n_cols = left_table.num_columns();
+    std::vector<std::unique_ptr<column>> merged_cols;
+    merged_cols.reserve(n_cols);
 
     column_merger merger{merged_indices, mr, stream};
+    transform(left_table.begin(), left_table.end(), 
+              right_table.begin(), 
+              std::back_inserter(merged_cols),
+              [&](auto const& left_col, auto const& right_col){
+                return cudf::experimental::type_dispatcher(left_col.type(),
+                                                           merger,
+                                                           left_col,
+                                                           right_col);
+              });
+
+    return std::make_unique<cudf::experimental::table>(std::move(merged_cols));
+}
 
-    for(auto i=0;i<n_cols;++i)
-      {
-        const auto& left_col = left_table.column(i);
-        const auto& right_col= right_table.column(i);
-
-        auto merged = cudf::experimental::type_dispatcher(left_col.type(),
-                                                          merger,
-                                                          left_col,
-                                                          right_col);
-        v_merged_cols.emplace_back(std::move(merged));
-      }
+struct merge_queue_item {
+  table_view view_;
+  table_ptr_type table_;
+  // Priority is a separate member to ensure that moving from an object 
+  // does not change its priority (which would ruin the queue invariant)
+  cudf::size_type priority_ = 0;
+ 
+  void set_priority() {
+    // Smallest tables have the highest priority
+    priority_ = -view_.num_rows();
+  }
+ 
+  merge_queue_item(table_view const& view, table_ptr_type&& table):
+  view_(view), 
+  table_(std::move(table)) {
+    set_priority();
+  }
+ 
+  bool operator<(merge_queue_item const& other) const {
+      return priority_ < other.priority_;
+  }
+};
 
-    return std::make_unique<cudf::experimental::table>(std::move(v_merged_cols));
+// Helper function to ensure that moving out of the priority_queue is "atomic"
+template <typename T>
+T top_and_pop(std::priority_queue<T>& q){
+  auto moved = std::move(const_cast<T&>(q.top()));
+  q.pop();
+  return moved;
 }
+ 
+table_ptr_type merge(std::vector<table_view> const& tables_to_merge,
+                     std::vector<cudf::size_type> const& key_cols,
+                     std::vector<cudf::order> const& column_order,
+                     std::vector<cudf::null_order> const& null_precedence,
+                     rmm::mr::device_memory_resource* mr,
+                     cudaStream_t stream = 0) {
+    if (tables_to_merge.empty()) {
+      return std::make_unique<cudf::experimental::table>();
+    }
+ 
+    auto const& first_table = tables_to_merge.front();
+    auto const n_cols = first_table.num_columns();
+ 
+    for (auto const& table: tables_to_merge) {
```

Comment:
Would it be clearer to just write these two conditions as separate `all` expectations?

```
CUDF_EXPECTS(std::all(tables_to_merge, [](auto const& table) { n_cols == table.num_columns(); }),
             "Mismatched number of columns");
```

A few times I've thought we should make a `CUDF_EXPECTS_ALL` macro that takes a container, a lambda/functor, and a message. But it only saves about 8 characters.

Reply:
I think it saves more than that, the code ended up as 
```
    CUDF_EXPECTS(std::all_of(tables_to_merge.cbegin(), tables_to_merge.cend(),
                             [n_cols] (auto const& tbl) { return n_cols == tbl.num_columns(); }),
                 "Mismatched number of columns");
```